### PR TITLE
Iteration of Interpolation Knots

### PR DESCRIPTION
**Pull request by awadell1**
# Overview of Change #
This pull request addresses Issue #346 by adding `knots(itp)` method to get a knots iterator for interpolations and extrapolations.

Support for the following has been implemented / tests (Please let me know if I missed something)
- Directional Extrapolation Boundary Conditions
- Nd Interpolations / Extrapolations (Iteration is handled on each axis separately and combined with Iterators.product)
- Non-equally spaced grid (i.e., GriddedInterpolation)

## Request for Comments ##
This would be my first public pull request, so here's hoping it's decent  :)

Currently, my test suite is passing (And hooked into the rest of the test suite), but if I missed a corner case or support for an interpolant, please let me know.

## Examples ##
The behavior for 1 and Nd interpolations essentially just spits out the results of getknots but flattened
```julia-repl
julia> using Interpolations;
julia> itp = interpolate(1:5, BSpline(Linear()));
julia> knots(itp) |> collect
5-element Array{Int64,1}:
 1
 2
 3
 4
 5

julia> itp = interpolate(rand(3,3), BSpline(Linear()));
julia> knots(itp) |> collect
3×3 Array{Tuple{Int64,Int64},2}:
 (1, 1)  (1, 2)  (1, 3)
 (2, 1)  (2, 2)  (2, 3)
 (3, 1)  (3, 2)  (3, 3)
```

Iteration over extrapolations will yield an infinite sequence of knots (For Periodic / Reflect) boundary conditions
```julia-repl
julia> x = [1.0, 2.3, 2.4, 2.8, 3.0];
julia> etp = LinearInterpolation(x, x.^2, extrapolation_bc=Periodic());
julia> knots(etp) |> x -> Iterators.take(x, 8) |> collect
8-element Array{Float64,1}:
 1.0
 2.3
 2.4
 2.8
 3.0
 4.3
 4.4
 4.8
```

## Next Steps (?) ##
I'm interested in a knot iterator like this as a stepping stone to implementing Gaussian Quadrature integration. The next step in that direction would be to add a `knotsbetween(itp; start, stop)` method to iterate over all knots between `start` and `stop` inclusively.

Is that a direction worth heading, or would that be out of scope for this package?

Best,

Alex

### CODE REVIEW

File: src/iterate.jl [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    nknots::Int
+    KnotIterator{T,ET}(k::AbstractArray{T}, bc::ET) where {T,ET} = new(k, bc, length(k))
+end
+
+KnotIterator(k::NTuple{N,AbstractArray}, bc::NTuple{N, ExtrapSpec}) where {N} = map(KnotIterator, k, bc)
+KnotIterator(k::NTuple{N,AbstractArray}, bc::BoundaryCondition) where {N} = map(x -> KnotIterator(x, bc), k)
+KnotIterator(k::AbstractArray{T}, bc::ET) where {T,ET <: ExtrapDimSpec} = KnotIterator{T,ET}(k, bc)
+
+const RepeatKnots = Union{Periodic,Reflect}
+Base.IteratorSize(::Type{KnotIterator}) = Base.SizeUnknown()
```
Comment (mkitti) [inline src/iterate.jl:24]: Rather than prefixing `Base` like this, perhaps we should explicitly import these in Interpolations.jl ?

File: src/iterate.jl [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+struct KnotIterator{T,ET <: ExtrapSpec}
+    knots::Vector{T}
+    bc::ET
+    nknots::Int
+    KnotIterator{T,ET}(k::AbstractArray{T}, bc::ET) where {T,ET} = new(k, bc, length(k))
+end
+
+KnotIterator(k::NTuple{N,AbstractArray}, bc::NTuple{N, ExtrapSpec}) where {N} = map(KnotIterator, k, bc)
+KnotIterator(k::NTuple{N,AbstractArray}, bc::BoundaryCondition) where {N} = map(x -> KnotIterator(x, bc), k)
+KnotIterator(k::AbstractArray{T}, bc::ET) where {T,ET <: ExtrapDimSpec} = KnotIterator{T,ET}(k, bc)
```
Comment (mkitti) [inline src/iterate.jl:21]: Add tests

File: src/iterate.jl [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+const RepeatKnots = Union{Periodic,Reflect}
+Base.IteratorSize(::Type{KnotIterator}) = Base.SizeUnknown()
+Base.IteratorSize(::Type{KnotIterator{T}}) where {T} = Base.SizeUnknown()
+
+Base.IteratorSize(::Type{KnotIterator{T,ET}}) where {T,ET} = _knot_iter_size(ET)
+_knot_iter_size(::Type{<:BoundaryCondition}) = Base.HasLength()
+_knot_iter_size(::Type{<:RepeatKnots}) = Base.IsInfinite()
+_knot_iter_size(::Type{Tuple{RevBC, FwdBC}}) where {RevBC,FwdBC} = _knot_iter_size(FwdBC)
```
Comment (mkitti) [inline src/iterate.jl:30]: Add tests

File: src/iterate.jl [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+Base.length(iter::KnotIterator) = _knot_length(iter, Base.IteratorSize(iter))
+_knot_length(iter::KnotIterator, ::Base.HasLength) = iter.nknots
+Base.size(iter::KnotIterator) = (length(iter),)
```
Comment (mkitti) [inline src/iterate.jl:34]: Add tests

File: src/iterate.jl [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+Base.IteratorEltype(::Type{KnotIterator{T,ET}}) where {T,ET} = Base.HasEltype()
+Base.eltype(::Type{KnotIterator{T,ET}}) where {T,ET} = T
```
Comment (mkitti) [inline src/iterate.jl:37]: Add tests

File: src/iterate.jl [hunk 6]
```diff
@@ -0,0 +1,138 @@
+
+export knots
```
Comment (mkitti) [inline src/iterate.jl:2]: Please move to Interpolations.jl file so the imports and exports are all together.

**Comment by mkitti** [comment]
Thank you for the pull request. The idea looks good. Ping me when the tests are passing please. Thanks.

Regarding `knotsbetween(itp; start, stop)` why should the arguments be keywords?

File: src/iterate.jl [hunk 7]
```diff
@@ -0,0 +1,149 @@
+# Similar to ExtrapDimSpec but for only a single dimension
+const ExtrapSpec = Union{BoundaryCondition,Tuple{BoundaryCondition,BoundaryCondition}}
+
+# Macro to get create ExtrapSpec for checking if a KnotIterator has a given BC
+# for forward iteration
+macro FwdExtrapSpec(bc)
+    :( Union{$bc,Tuple{BoundaryCondition,$bc}} )
+end
+struct KnotIterator{T,ET <: ExtrapSpec}
```
Comment (mkitti) [inline src/iterate.jl:9]: Add proper doc string for `KnotIterator`

File: src/iterate.jl [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    iter = KnotIterator(k, bc)
+    length(iter) == 1 ? iter[1] : Iterators.product(iter...)
+end
+
+# For non-repeating ET's iterate through once
+iterate(iter::KnotIterator) where {T} = iterate(iter, 1)
+iterate(iter::KnotIterator, idx::Integer) = idx <= iter.nknots ? (iter.knots[idx], idx+1) : nothing
+
+# For repeating knots state is the knot index + offset value
+function iterate(iter::KnotIterator{T,ET}) where {T,ET <: @FwdExtrapSpec RepeatKnots}
```
Comment (mkitti) [inline src/iterate.jl:98]: Codecov is saying this the only line left not covered by tests. I'm not sure why though.

**Review by mkitti — APPROVED** [review]
Great. This looks like it is almost mergeable. The only thing left is to shore up the documentation for the `KnotIterator` struct type and making the example code into a jldoctest per Documenter.jl.

A markdown file in the docs directory as well into the Documenter.jl docs would be appreciated.

**Comment by awadell1** [comment]
Matt,

Yeah, not sure what's going on with the lack of coverage for that line. It has to get hit otherwise, `state` doesn't get initialized correctly. Plus line 99 gets hit, so 98 must get hit too?

> Regarding `knotsbetween(itp; start, stop)` why should the arguments be keywords?

So my thought here was to mimic the syntax of range: `range(start; length::Union{Integer,Nothing}=nothing, stop=nothing, step=nothing)`.

So the syntax would be:

- `knotsbetween(itp; start=x)` Iterates over all knots greater than x
- `knotsbetween(itp; stop=x)` Iterates over all knots less than x
- `knotsbetween(itp; start=x, stop=y)` Iterates over all knots between x and y

Although maybe that should be rolled into `knots`?

Thank you for the fast review, I'm working through your last round of comments, and will ping you when it's done.

Best,

Alex

**Comment by awadell1** [comment]
Locally I can show full coverage for iterate.jl if I run julia with inlining disabled:

```shell
julia --inline=no --project -e '\
		using Pkg, Coverage; \
		Pkg.test("Interpolations", coverage=true); \
		LCOV.writefile("lcov.info", process_folder());'
```

Doctests wise, it looks like everything is passing now. Do you have a script you use to build the docs locally? I think I have everything working correctly locally, aside from the following warnings:

```shell
julia --project make.jl
[ Info: SetupBuildDirectory: setting up build directory.
[ Info: Doctest: running doctests.
[ Info: ExpandTemplates: expanding markdown templates.
[ Info: CrossReferences: building cross-references.
[ Info: CheckDocument: running document checks.
┌ Warning: 1 docstring potentially missing:
│ 
│     Interpolations.l4_2d_cs
└ @ Documenter.DocChecks ~/.julia/packages/Documenter/PLD7m/src/DocChecks.jl:63
[ Info: Populate: populating indices.
[ Info: RenderDocument: rendering document.
[ Info: HTMLWriter: rendering HTML pages.
┌ Warning: invalid local image: unresolved path in control.md
│   link = "doc/images/parametric_spline.png"
└ @ Documenter.Writers.HTMLWriter ~/.julia/packages/Documenter/PLD7m/src/Writers/HTMLWriter.jl:1718
┌ Warning: Documenter could not auto-detect the building environment Skipping deployment.
└ @ Documenter ~/.julia/packages/Documenter/PLD7m/src/deployconfig.jl:41

**Comment by mkitti** [comment]
There are some other issues we need to fix with the documentation, but otherwise this looks good. If you're done, I'll merge this. Thanks. -Mark

**Comment by awadell1** [comment]
Yeah, I'm done with the `knots` work. I'll open a separate pull request for adding `knotsbetween` when that's ready.

Best,
Alex
